File: tests/cpu/python/test_timer.py
"""Timer tests for Insight7 Python binding."""
import time
import insight as ins


def test_timer_cpu_create():
    t = ins.Timer((0, 0))
    assert t is not None
    assert hasattr(t, "start")
    assert hasattr(t, "stop")
    assert hasattr(t, "elapsed_ms")


def test_timer_cpu_measure():
    t = ins.Timer((0, 0))
    t.start()
    time.sleep(0.005)
    t.stop()
    ms = t.elapsed_ms()
    assert ms >= 0.0
    assert ms < 100.0  # Should be ~5ms


def test_timer_cpu_context_manager():
    with ins.Timer((0, 0)) as t:
        time.sleep(0.01)
    ms = t.elapsed_ms()
    assert ms >= 0.0


def test_timer_cpu_manual():
    t = ins.Timer((0, 0))
    t.start()
    time.sleep(0.003)
    t.stop()
    ms = t.elapsed_ms()
    assert ms >= 0.0


def test_timer_invalid_place():
    import pytest
    with pytest.raises(TypeError):
        ins.Timer("invalid")
    with pytest.raises(TypeError):
        ins.Timer(123)


def test_timer_gpu():
    if not ins.is_device_available("gpu"):
        pytest.skip("GPU not available")
    import insight as ins
    ins.load_backend("cuda")
    t = ins.Timer((1, 0))
    t.start()
    a = ins.ones([256, 256], dtype=ins.float32, place=ins.GPUPlace(0))
    b = ins.full([256, 256], 2.0, dtype=ins.float32, place=ins.GPUPlace(0))
    c = a + b
    t.stop()
    ms = t.elapsed_ms()
    assert ms >= 0.0

File: tests/cpu/lua/test_timer.lua
-- test_timer.lua — Timer tests for Insight7 Lua binding
local ins = require("insight")

describe("Timer (CPU)", function()
  it("creates and destroys timer", function()
    local t = ins.Timer(0, 0)
    assert.is_not_nil(t)
    assert.is_not_nil(t._handle)
  end)

  it("measures elapsed time", function()
    local t = ins.Timer(0, 0)
    t:start()
    -- Busy wait for ~10ms
    local start = os.clock()
    while os.clock() - start < 0.01 do
    end
    t:stop()
    local ms = t:elapsed()
    assert.is_true(ms >= 0.0)
    assert.is_true(ms < 100.0)
  end)

  it("supports manual start/stop", function()
    local t = ins.Timer(0, 0)
    t:start()
    local start = os.clock()
    while os.clock() - start < 0.005 do
    end
    t:stop()
    local ms = t:elapsed()
    assert.is_true(ms >= 0.0)
  end)

  it("raises error on nil arguments", function()
    assert.has_error(function()
      ins.Timer(nil, nil)
    end)
    assert.has_error(function()
      ins.Timer(0, nil)
    end)
  end)

  it("raises error on non-numeric arguments", function()
    assert.has_error(function()
      ins.Timer("cpu", "0")
    end)
  end)
end)
